homeassistant/helpers/translation.py

```diff
@@ -465,30 +452,62 @@ def async_setup(hass: HomeAssistant) -> None:
 
     Listeners load translations for every loaded component and after config change.
     """
+    cache = _TranslationCache(hass)
 
-    hass.data[TRANSLATION_FLATTEN_CACHE] = _TranslationCache(hass)
+    hass.data[TRANSLATION_FLATTEN_CACHE] = cache
 
-    async def load_translations(event: Event) -> None:
-        if "language" in event.data:
-            language = hass.config.language
-            _LOGGER.debug("Loading translations for language: %s", language)
-            await _async_load_state_translations_to_cache(hass, language, None)
+    @callback
+    def _async_load_translations_filter(event: Event) -> bool:
+        """Filter out unwanted events."""
+        return "language" in event.data
 
-    async def load_translations_for_component(event: Event) -> None:
-        component = event.data.get("component")
+    async def _async_load_translations(event: Event) -> None:
+        """Load translations for the current language."""
+        language = hass.config.language
+        _LOGGER.debug("Loading translations for language: %s", language)
+        await cache.async_load(language, hass.config.components)
+
+    @callback
+    def _async_load_translations_for_component_filter(event: Event) -> bool:
+        """Filter out unwanted events."""
+        component: str | None = event.data.get("component")
         # Platforms don't have their own translations, skip them
-        if component is None or "." in str(component):
-            return
+        return bool(
+            component
+            and "." not in component
+            and not cache.async_is_loaded(hass.config.language, {component})
+        )
+
+    async def _async_load_translations_for_component(event: Event) -> None:
+        """Load translations for a component."""
+        component: str | None = event.data.get("component")
+        if TYPE_CHECKING:
+            assert component is not None
         language = hass.config.language
         _LOGGER.debug(
             "Loading translations for language: %s and component: %s",
-            hass.config.language,
+            language,
             component,
         )
-        await _async_load_state_translations_to_cache(hass, language, component)
+        await cache.async_load(language, {component})
```

Comment:
This call will almost never happen now since it will already be loaded unless the integration is loaded late